sockeye/data_io.py

```diff
@@ -312,30 +292,48 @@ def tokens2ids(tokens: Iterable[str], vocab: Dict[str, int]) -> List[int]:
     return [vocab.get(w, vocab[C.UNK_SYMBOL]) for w in tokens]
 
 
-def read_sentences(path: str, vocab: Dict[str, int], add_bos=False, limit=None) -> List[List[int]]:
+class SentenceReader(Iterator):
     """
-    Reads sentences from path and creates word id sentences.
+    Reads sentences from patch and creates word id sentences.
+    Streams from disk, instead of loading all sentences into memory.
 
     :param path: Path to read data from.
     :param vocab: Vocabulary mapping.
     :param add_bos: Whether to add Beginning-Of-Sentence (BOS) symbol.
     :param limit: Read limit.
-    :return: List of integer sequences.
     """
-    assert C.UNK_SYMBOL in vocab
-    assert C.UNK_SYMBOL in vocab
-    assert vocab[C.PAD_SYMBOL] == C.PAD_ID
-    assert C.BOS_SYMBOL in vocab
-    assert C.EOS_SYMBOL in vocab
-    sentences = []
-    for sentence_tokens in read_content(path, limit):
-        sentence = tokens2ids(sentence_tokens, vocab)
-        check_condition(bool(sentence), "Empty sentence in file %s" % path)
-        if add_bos:
-            sentence.insert(0, vocab[C.BOS_SYMBOL])
-        sentences.append(sentence)
-    logger.info("%d sentences loaded from '%s'", len(sentences), path)
-    return sentences
+
+    def __init__(self, path: str, vocab: Dict[str, int], add_bos: bool = False, limit: Optional[int] = None) -> None:
+        self.path = path
+        self.vocab = vocab
+        self.add_bos = add_bos
+        self.limit = limit
+        assert C.UNK_SYMBOL in vocab
+        assert C.UNK_SYMBOL in vocab
+        assert vocab[C.PAD_SYMBOL] == C.PAD_ID
+        assert C.BOS_SYMBOL in vocab
+        assert C.EOS_SYMBOL in vocab
+        self._iter = None  # type: Optional[Iterator]
+        self._iterated_once = False
+        self.count = 0
+
+    def __iter__(self):
+        self._iter = read_content(self.path, self.limit)
```

Comment:
I'm not a huge fan of this as the following will not do what you'd expect:
```
iter1 = iter(sentence_reader)
el1 = iter1.next()
iter2 = iter(sentence_reader) # resets the iter
el2 = iter1.next() # gives the first element a second time.
```
how about raising `if self._iter is not None` to make it explicit that this kind of usage is forbidden?

Reply:
oh good catch, I will add a check!